### PR TITLE
Distribute judge-tied non-priority legacy hearings first

### LINKED ISSUE #14542: Update case orders logic
Update case distribution order to:
1. Reduce backlog of legacy hearing cases tied to the judge to 30
1. Continue with current algorithm

<details><summary> Newer old requirements </summary>

User Story: When a Veteran Law Judge requests more cases, the case distribution logic should be as follows to ensure cases are worked in the correct order: 

1. Distribute non-priority cases tied to the judge
    a. Legacy hearing cases tied to the judge - if the judge has more than 30 tied to them, fill their case order with only legacy hearing cases tied to them until the threshold of 30 has been hit or until their case order (batch size) has been met,  whichever comes first.  If the judge has <= 30 legacy hearing cases tied to them continue with the below. Creation of backlog [here](https://github.com/department-of-veterans-affairs/caseflow/issues/14543)
    b. AMA hearing cases with affinity for the judge
    c. Distribute non-priority cases that are general population based on docket proportions using stochastic rounding

We no longer want Priority appeals to be distributed via the Request Cases logic, but instead be automatically distributed when they are ready #14604 

## Current
We limit the number of priority appeals distributed in a distribution by ensuring we do not distribute more priority appeals than our priority target. [Priority target is calculated](https://github.com/department-of-veterans-affairs/caseflow/blob/5bbbe90117365a8689e3f894a3cc4223ea229569/app/models/concerns/ama_case_distribution.rb#L112) as the ratio of [sum of priority cases on all dockets](https://github.com/department-of-veterans-affairs/caseflow/blob/57e746e91b91c82c9b5f4edac343bde9688b68d9/app/models/docket_coordinator.rb#L88-L93) to [the total number of attorneys in caseflow * 3](https://github.com/department-of-veterans-affairs/caseflow/blob/57e746e91b91c82c9b5f4edac343bde9688b68d9/app/models/docket_coordinator.rb#L123-L125). This has lead to a steady increase in the number of priority appeals waiting to be distributed (`Distribution.statistics.priority_count` in [Nov 2018](http://query.prod.appeals.va.gov/question#eyJuYW1lIjpudWxsLCJkYXRhc2V0X3F1ZXJ5Ijp7ImRhdGFiYXNlIjoyLCJ0eXBlIjoicXVlcnkiLCJxdWVyeSI6eyJzb3VyY2UtdGFibGUiOjE1LCJvcmRlci1ieSI6W1siYXNjIixbImZpZWxkLWlkIiwyOTZdXV19fSwiZGlzcGxheSI6InRhYmxlIiwidmlzdWFsaXphdGlvbl9zZXR0aW5ncyI6e319) vs [now](http://query.prod.appeals.va.gov/question#eyJuYW1lIjpudWxsLCJkYXRhc2V0X3F1ZXJ5Ijp7ImRhdGFiYXNlIjoyLCJ0eXBlIjoicXVlcnkiLCJxdWVyeSI6eyJzb3VyY2UtdGFibGUiOjE1LCJvcmRlci1ieSI6W1siZGVzYyIsWyJmaWVsZC1pZCIsMjk2XV1dfX0sImRpc3BsYXkiOiJ0YWJsZSIsInZpc3VhbGl6YXRpb25fc2V0dGluZ3MiOnt9fQ==)).

We also fill a distribution with legacy and ama hearing docket cases with [non priority cases associated with the judge](https://github.com/department-of-veterans-affairs/caseflow/blob/5bbbe90117365a8689e3f894a3cc4223ea229569/app/models/concerns/ama_case_distribution.rb#L30-L33) before we [distribute priority cases that are not associated with the judge](https://github.com/department-of-veterans-affairs/caseflow/blob/5bbbe90117365a8689e3f894a3cc4223ea229569/app/models/concerns/ama_case_distribution.rb#L38). 

<details><summary>Old requirements, now outdated</summary>
After collecting stats on how the removal of this priority factor will affect other docket distributions and the go-ahead from the board, remove this priority target by replacing `priority_rem` with `@rem`.~

```diff
-     # If we haven't yet met the priority target, distribute additional priority appeals.
-     priority_rem = (priority_target - @appeals.count(&:priority)).clamp(0, @rem)
-     oldest_priority_appeals_by_docket(priority_rem).each do |docket, n|
-       distribute_appeals(docket, n, priority: true)
-     end
+     # If we haven't yet filled the request order, distribute additional priority appeals.
+     oldest_priority_appeals_by_docket(@rem).each do |docket, n|
+       distribute_appeals(docket, n, priority: true)
+     end
```
The priority cases that are **not** associated with the judge should take precedence over the **non-priority** cases associated with the judge, so ensure we move this distribution to before we  [distribute **non-priority** cases that are associated with the judge](https://github.com/department-of-veterans-affairs/caseflow/blob/5bbbe90117365a8689e3f894a3cc4223ea229569/app/models/concerns/ama_case_distribution.rb#L30-L33), but after we [distribute **priority** cases that are associated with the judge](https://github.com/department-of-veterans-affairs/caseflow/blob/5bbbe90117365a8689e3f894a3cc4223ea229569/app/models/concerns/ama_case_distribution.rb#L26-L28)
</details>
</details>

## Acceptance Criteria
 - [ ] Please put this work behind the feature toggle: Priority Case Auto-Distribution
 - [ ] Every case order (distribution request) is filled with legacy hearing cases tied to the judge first. If the judge has more than 30 tied to them distribute their legacy hearing cases until the threshold of 30 has been hit or until their case order (batch size) has been met, whichever comes first
 - [ ] Once the judge is down to 30 cases in their backlog, if they have more cases to fill in their request, fill the case order per the original algorithm rules

### To reduce confusion
Move these 3 lines https://github.com/department-of-veterans-affairs/caseflow/blob/6287312416efb2e55988a3d1bf5d69fbfefb235d/app/models/concerns/ama_case_distribution.rb#L32-L34

to before these three lines. https://github.com/department-of-veterans-affairs/caseflow/blob/6287312416efb2e55988a3d1bf5d69fbfefb235d/app/models/concerns/ama_case_distribution.rb#L26-L28

No other changes, do not remove priority

### PR DESCRIPTION
**Pull request by yoomlam**
Resolves #14542 

### Description
Distribute judge-tied legacy hearings first when distributing cases in order to reduce the backlog of legacy hearing cases tied to the judge down to 30.

### Acceptance Criteria
 - [ ] this work is behind the feature toggle: `priority_acd`
 - [ ] Every case order (distribution request) is filled with legacy hearing cases tied to the judge first. If the judge has more than 30 tied to them distribute their legacy hearing cases until the threshold of 30 has been hit or until their case order (batch size) has been met, whichever comes first
 - [ ] Once the judge is down to 30 cases in their backlog, if they have more cases to fill in their request, fill the case order per the original algorithm rules

### Testing Plan
See https://github.com/department-of-veterans-affairs/caseflow/pull/14615#issue-441503304

### CODE REVIEW

File: app/models/concerns/ama_case_distribution.rb [hunk 1]
```diff
@@ -23,16 +23,16 @@ def ama_distribution
...
     # Distribute legacy cases tied to a judge down to the board provided limit of 30, regardless of the legacy docket
     # range
     if FeatureToggle.enabled?(:priority_acd)
       distribute_appeals(:legacy, @rem, priority: false, genpop: "not_genpop", bust_backlog: true)
     end
 
+    # Distribute priority appeals that are tied to judges (not genpop).
+    distribute_appeals(:legacy, @rem, priority: true, genpop: "not_genpop")
+    distribute_appeals(:hearing, @rem, priority: true, genpop: "not_genpop")
+
```
Comment (yoomlam) [inline app/models/concerns/ama_case_distribution.rb:35]: Reorder as Hunter instructed in https://github.com/department-of-veterans-affairs/caseflow/issues/14542#issue-639909197

File: spec/models/distribution_spec.rb [hunk 2]
```diff
@@ -34,71 +34,49 @@
...
-          bfdloout: i.days.ago,
-          folder: build(
-            :folder,
-            tinum: "1801#{format('%<index>03d', index: i)}",
-            titrnum: "123456789S"
-          )
-        )
-      end
+      (15..100).map { |i| create_legacy_case(i) }
+    end
```
Comment (yoomlam) [inline spec/models/distribution_spec.rb:60]: DRYing it up part 1.

File: spec/models/distribution_spec.rb [hunk 3]
```diff
@@ -34,71 +34,49 @@
...
 
     let!(:other_judge_hearings) do
-      legacy_nonpriority_cases[2..27].map do |appeal|
-        create(:case_hearing,
-               :disposition_held,
-               folder_nr: appeal.bfkey,
-               hearing_date: 1.month.ago,
-               board_member: "1234")
-      end
+      legacy_nonpriority_cases[2..27].map { |appeal| create_legacy_case_hearing_for(appeal, board_member: "1234") }
```
Comment (yoomlam) [inline spec/models/distribution_spec.rb:79]: DRYing it up part 2.

File: spec/models/distribution_spec.rb [hunk 4]
```diff
@@ -203,27 +181,63 @@
...
+
+          # distributions after handling BACKLOG_LIMIT
+          remaining_in_backlog = total_tied_nonpriority_hearings -
+                                 dcs_legacy.where(priority: false, genpop: false).count
+          expect(subject.statistics["legacy_hearing_backlog_count"]).to eq remaining_in_backlog
+
+          expect(dcs_legacy.where(priority: true, genpop_query: "not_genpop").count).to eq(2)
+          expect(dcs_legacy.where(priority: true, genpop_query: "any").count).to eq(2)
+          expect(dcs_legacy.count).to be >= 8
+        end
```
Comment (yoomlam) [inline spec/models/distribution_spec.rb:240]: Added 1 test here. The other changes are refactors.

File: spec/models/distribution_spec.rb (hunk 2 above)
Comment (hschallhorn) [reply spec/models/distribution_spec.rb:60]: Perfect!

**Pull request merged**